### PR TITLE
issue #1777 - extract compartment inclusion params even when filtered

### PR DESCRIPTION
**Pull request by lmsurpre**
1. When building our ParametersMaps (e.g. on server start), use CompartmentUtil to get the list of compartment inclusion criteria params (e.g. `patient` for the Observation resourceType on the Patient compartment) and add them to a new Map within ParametersMap (separate from the other search parameters).
   * If the corresponding inclusion criteria param is explicitly configured in the config, use that
   * If not, then look it up in the registry and log a warning if there is more than one SearchParameter for that Resource+code (just like normal ParametersMap loading).

2. When extracting search parameter values in SearchUtil, use CompartmentUtil to get the list of compartment inclusion criteria params. For those that aren't configured as external search parameters, add the inclusion parameter from this special map in ParametersMap and add a special DO_NOT_STORE extension to differentiate them from the external search parameters.
    
3. When processing the search parameters in FHIRPersistenceJDBCImpl, propagate the DO_NOT_STORE extension into the ExtractedParameterValues via a new "isForStoring" member variable.  Then, use CompartmentUtil to get the list of compartment inclusion criteria params one last time and process the corresponding extracted parameter values.
   * If "isForStoring", then this is an external search parameter and so we make a copy of the ExtractedParameterValue for our `ibm-internal-[x]-compartment` param and add it to the list
   * If not "isForStoring", then rename the search parameter to `ibm-internal-[x]-compartment` in-place and set isForStoring to true
   
4. Added a quickfix for #3283 and also updated corresponding fhir-search tests to debug some failing tests and, eventually,  work around a design limitation I found in ParametersUtil (and opened #3385 for that)

### CODE REVIEW

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 1]
```diff
@@ -490,20 +488,6 @@ private void doCachePrefill() throws FHIRPersistenceException {
...
-     * Creates and returns a copy of the passed resource with the {@code Resource.id}
-     * {@code Resource.meta.versionId}, and {@code Resource.meta.lastUpdated} elements replaced.
-     *
-     * @param resource
-     * @param logicalId
-     * @param newVersionNumber
-     * @param lastUpdated
-     * @return the updated resource
-     */
-    private <T extends Resource> T copyAndSetResourceMetaFields(T resource, String logicalId, int newVersionNumber, Instant lastUpdated) {
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:503]: unrelated cleanup of unused private helper

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 2]
```diff
@@ -223,7 +221,7 @@
 
     // Enable use of legacy whole-system search parameters for the search request
     private final boolean legacyWholeSystemSearchParamsEnabled;
-    
+
```
Comment (lmsurpre) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:224]: lots of whitespace changes on this file...to focus on actual changes you may want to hide the whitespace ones

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 3]
```diff
@@ -1800,39 +1795,67 @@ private boolean isWholeSystem(SearchParameter sp) {
...
+                }
 
-                if (log.isLoggable(Level.FINE)) {
-                    log.fine("Adding compartment reference parameter: [" + resourceType + "] "+ parameterName + " = " + rv.getTargetResourceType() + "/" + rv.getValue());
+                ReferenceValue rv = ((ReferenceParmVal) epv).getRefValue();
+                if (rv != null && rv.getType() == ReferenceType.LITERAL_RELATIVE
+                        && compartments.contains(rv.getTargetResourceType())) {
+                    String internalCompartmentParamName = CompartmentUtil.makeCompartmentParamName(rv.getTargetResourceType());
+
+                    if (epv.isForStoring()) {
```
Comment (tbieste) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:1840]: The code looks good.  I just don't quite follow the "isForStoring" meaning.  Store where, etc?  I think I just need some extra education on the side for background.

**Review by tbieste — COMMENTED** [review]
It would be good to add some testcases in some fashion for this.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 3 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:1840]: for storing in the search parameter tables in our db

**Comment by lmsurpre** [comment]
> It would be good to add some testcases in some fashion for this.

I just added a couple unit tests and an e2e test under SearchTest.  See if you think thats adequate or if I should be adding some more.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 3 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:1840]: basically this was just the cleanest way I could think of the get some parameters extracted from our existing SearchUtil.extractParameters method without actually paying the cost of storing them in the db

File: fhir-model/src/test/java/com/ibm/fhir/model/test/TestUtil.java [hunk 4]
```diff
@@ -117,12 +116,26 @@ public static JsonObject getRequestJsonObject(String method, String url) {
...
             .toBuilder()
-            .subject(Reference.builder().reference(string("Patient/" + patientId)).build())
+            .subject(Reference.builder().reference("Patient/" + patientId).build())
+            .build();
+        return observation;
+    }
+
+    /**
+     * Loads an Observation resource from the specified file, then associates it with
+     * the specified patient via a subject attribute.
```
Comment (tbieste) [inline fhir-model/src/test/java/com/ibm/fhir/model/test/TestUtil.java:130]: Also mention it sets the Practitioner / performer.

File: fhir-search/src/test/java/com/ibm/fhir/search/parameters/ParametersUtilTest.java [hunk 5]
```diff
@@ -49,7 +49,7 @@ public void testGetAllSearchParameters() throws IOException {
             ParametersUtil.print(out);
```
Comment (tbieste) [inline fhir-search/src/test/java/com/ibm/fhir/search/parameters/ParametersUtilTest.java:49]: Update copyright date.

File: fhir-model/src/test/java/com/ibm/fhir/model/test/TestUtil.java (hunk 4 above)
Comment (lmsurpre) [reply fhir-model/src/test/java/com/ibm/fhir/model/test/TestUtil.java]: done

File: fhir-search/src/test/java/com/ibm/fhir/search/parameters/ParametersUtilTest.java (hunk 5 above)
Comment (lmsurpre) [reply fhir-search/src/test/java/com/ibm/fhir/search/parameters/ParametersUtilTest.java:49]: done

**Review by tbieste — APPROVED** [review]

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 2 above)
Comment (punktilious) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:224]: could end up with some merge issues with the payload offloading PR that's pending

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 6]
```diff
@@ -1739,14 +1729,19 @@ private ExtractedSearchParameters extractSearchParameters(Resource fhirResource,
...
                 // If this is a definitional resource, augment the extracted parameter list with a composite
                 // parameter that will be used for canonical searches. It will contain the url and version
                 // values.
                 addCanonicalCompositeParam(allParameters);
             }
 
+            // Remove parameters that aren't to be stored
+            boolean anyRemoved = allParameters.removeIf(value -> !value.isForStoring());
+            if (anyRemoved) {
+                log.warning("Removed unexpectedly found extracted search parameter values that weren't for storing");
```
Comment (punktilious) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:1743]: warning text is a little confusing

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java [hunk 7]
```diff
@@ -1739,14 +1729,19 @@ private ExtractedSearchParameters extractSearchParameters(Resource fhirResource,
...
+                addCompartmentParams(allParameters, fhirResource.getClass().getSimpleName());
 
                 // If this is a definitional resource, augment the extracted parameter list with a composite
                 // parameter that will be used for canonical searches. It will contain the url and version
                 // values.
                 addCanonicalCompositeParam(allParameters);
             }
 
+            // Remove parameters that aren't to be stored
+            boolean anyRemoved = allParameters.removeIf(value -> !value.isForStoring());
```
Comment (punktilious) [inline fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:1741]: removeIf isn't all that cheap on an ArrayList. Might be worth considering a linked list for allParameters - which has much lower remove costs.

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 6 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:1743]: we should never hit it anyway.  open to suggestion

Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:1743]: is this better?
```suggestion
                log.warning("The set of extracted parameters values unexpectedly contained values that weren't for storing; these have been removed");
```

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/ParametersMap.java [hunk 8]
```diff
@@ -117,6 +139,10 @@ public SearchParameter lookupByCanonical(String searchParameterCanonical) {
         return canonicalMap.get(searchParameterCanonical);
     }
 
+    public SearchParameter getInclusionParam(String searchParameterCode) {
```
Comment (punktilious) [inline fhir-search/src/main/java/com/ibm/fhir/search/parameters/ParametersMap.java:142]: javadoc

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/ParametersMap.java [hunk 9]
```diff
@@ -137,6 +163,14 @@ public int size() {
         return Collections.unmodifiableSet(codeMap.entrySet());
     }
 
+    public Collection<SearchParameter> inclusionValues() {
```
Comment (punktilious) [inline fhir-search/src/main/java/com/ibm/fhir/search/parameters/ParametersMap.java:166]: javadoc

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 7 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:1743]: i made the change

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/ParametersMap.java (hunk 8 above)
Comment (lmsurpre) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/ParametersMap.java:158]: added

File: fhir-search/src/main/java/com/ibm/fhir/search/parameters/ParametersMap.java (hunk 9 above)
Comment (lmsurpre) [reply fhir-search/src/main/java/com/ibm/fhir/search/parameters/ParametersMap.java:190]: done

File: fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java (hunk 2 above)
Comment (lmsurpre) [reply fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java:225]: yeah, i need to go review that one.  if you want, we can hold on this one til thats merged...i don't mind handling conflicts
